### ISSUE TITLE
Unable to compose and send messages (504 Gateway Timeout)

### ISSUE THREAD

**Issue by AlistairTB** [roundcube/roundcubemail#8187]
OS: CentOS 7
PHP  7.3.29
RCM 1.4.11
MariaDB 10.6.4

To reproduce:
Open "compose message". fill in fields, hit send.

What happens: 
Message "Sending message" is displayed. If I open "Inspect Object" during this time, I eventually see "504 Gateway timeout" in the network messages. My web server and proxy server never see the request to post the message.

All other features of RCM work as expected. I can update profile settings, add concacts, etc.

This happens across multiple machines and browsers, specifically vivaldi and firefox.

Special considerations:
This started happening immediately after an upgrade from 1.3*ish to 1.4.11.

During the upgrade (using installto), I did encounter one error (screen shot), but it seemed to not be a big deal since everything seemed to work just fine after the upgrade. The error was:

```
"ERROR: Error in DDL upgrade 2018122300: [1205] Error on rename of '/rcm/filestore' to './rcm/#sql-backup-2e51-808fb' (erno: 168 "Unknown (generic) error from engine")
Updating database scheme (2018122300)... [FAILED]
All done.
```

Attaching about screenshot
![Screenshot from 2021-08-29 08-36-52](https://user-images.githubusercontent.com/89739791/131256332-57187ac0-3fbf-4a04-8e33-e6daff5b50fa.png)

**Comment by alecpl** [comment]
Duplicate of #8172? Does it work with all plugins disabled?

**Comment by AlistairTB** [comment]
I've disabled all plugins in config/config.inc.php and still encounter the problem.
I do not have any spell check options enabled that I'm aware of (Settings | Composing Messages). Screenshot attached. If there is another location that enables spell check options, please let me know so I can disable them fully.

![Screenshot from 2021-08-29 11-44-49](https://user-images.githubusercontent.com/89739791/131262174-ea27aae8-8cf3-4030-ae26-8ee41e307108.png)

![Screenshot from 2021-08-29 11-54-54](https://user-images.githubusercontent.com/89739791/131262165-e08932a2-be9b-4cf9-8c41-03c61a8efee0.png)

**Comment by AlistairTB** [comment]
So here is a snippet of httpd/apache logs
`10.11.80.13 - - [29/Aug/2021:12:43:03 -0700] "POST /mail/?_task=mail&_action=autocomplete HTTP/1.1" 200 1070 "https://example.com/mail/?_task=mail&_action=compose&_id=437789290612be33b6f140" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.126 Safari/537.36"
10.11.80.13 - - [29/Aug/2021:12:43:04 -0700] "POST /mail/?_task=mail&_action=autocomplete HTTP/1.1" 200 977 "https://example.com/mail/?_task=mail&_action=compose&_id=437789290612be33b6f140" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.126 Safari/537.36"

<< I HIT SEND HERE FROM THE COMPOSE WINDOW IN MY BROWSER. NO NEW ENTRIES SHOW UP. >>

10.11.80.13 - - [29/Aug/2021:12:43:46 -0700] "POST /mail/?_task=mail&_action=refresh HTTP/1.1" 200 984 "https://example.com/mail/?_task=mail&_mbox=INBOX" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36"`

This is what I copy from the developer toolbox when I try to send the message:
`Request URL: https://example.com/mail/?_task=mail&_unlock=loading1630243920278&_framed=1&_lang=en_US
Request Method: POST
Status Code: 504 Gateway Time-out
Remote Address: xxx.xxx.xxx.xxx:443
Referrer Policy: same-origin
Cache-Control: no-cache
Connection: close
Content-Type: text/html
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cache-Control: max-age=0
Connection: keep-alive
Content-Length: 324
Content-Type: application/x-www-form-urlencoded
Cookie: language=en_US; 09cc6cbe29e9df86b825ae49fc1d107b=lr2RNvDdtru4SAlrgO82mtkbwVg5ws2JzcpG143Qx3KQCy8xXrjUy6DRVleMlj-m7DlTbwMOQZpZgQsKh83S-1YU41v-Mg0cdgnGXCPkDtjVqCscW9aXSrydP1R05hF-; roundcube_sessid=lbnthtn66m48rvcvpj7k58av4j; i18next=en-US; roundcube_sessauth=xOMLWfsIgJWpWz6nSy9QYgBd93-1630243800
Host: example.com
Origin: https://example.com
Referer: https://example.com/mail/?_task=mail&_action=compose&_id=1736332215612b8c370645c&_extwin=1
sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="92"
sec-ch-ua-mobile: ?0
Sec-Fetch-Dest: iframe
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: same-origin
Sec-Fetch-User: ?1
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.126 Safari/537.36
_task: mail
_unlock: loading1630243920278
_framed: 1
_lang: en_US
_token: eTMVtDw1H34wgsSdHOvxpJdKsTnRNV54
_extwin: 1
_task: mail
_action: send
_id: 1736332215612b8c370645c
_attachments: 
_from: 106
_to: user@example.com,
_cc: 
_bcc: 
_replyto: 
_followupto: 
_subject: asldf
_draft_saveid: 
_draft: 
_is_html: 0
_framed: 1
_message: asdfasdf
editorSelector: plain
_mdn: 
_dsn: 
_priority: 0
_store_target: Sent`

I can't explain why this POST doesn't show up in my server logs and I'm at the end of my troubleshooting skills. Any tips to help me gather more information would be greatly appreciated.

**Comment by AlistairTB** [comment]
I can definitely see my haproxy sending the request to the backend apache server now, but the request is still never showing up in my logs. 
`14:39:58.017571 IP 10.11.80.13.35498 > 10.11.80.12.80: Flags [P.], seq 117587787:117589345, ack 1088789171, win 229, options [nop,nop,TS val 879502053 ecr 101027108], length 1558: HTTP: POST /mail/?_task=mail&_**unlock=loading1630273197870**&_framed=1&_lang=en_US HTTP/1.1`

My apache log eventually displays the expected/corresponding line in the log file, but only after the client displays the 504 gateway timeout. I know this is the related apache log entry because of the matching "unlock" number in the URI from the browser debug:

`10.11.80.13 - - [29/Aug/2021:14:58:46 -0700] "POST /mail/?_task=mail&_**unlock=loading1630273197870**&_framed=1&_lang=en_US HTTP/1.1" 200 648 "https://example.com/mail/?_task=mail&_action=compose&_id=735130496612c0301cbe07" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.126 Safari/537.36"
`
So it appears RCM is trying to process the incoming message, timesout, and the client gives up.

**Comment by AlistairTB** [comment]
I've added debugging to the RCM config in config/config.inc.php
$config['debug_level'] = 9;

Where do the logs get written to? There's nothing in <rcm_install>/logs directory. 
`======================
= Personal Web Server
======================
2021-08-29:1512 PDT
[root@web03 logs]# ls -lahZ
drwxrwxr-x. root apache unconfined_u:object_r:httpd_sys_rw_content_t:s0 .
drwxr-xr-x. root apache unconfined_u:object_r:httpd_sys_content_t:s0 ..
-rw-r--r--. root apache unconfined_u:object_r:httpd_sys_rw_content_t:s0 .htaccess
`

**Comment by AlistairTB** [comment]
Was missing several config options for debugging:
`
$config['debug_level'] = 9;
$config['log_dir'] = 'logs/';
$config['log_driver'] = 'file';
$config['log_date_format'] = 'd-M-Y H:i:s O';
$config['log_session_id'] = 0;
$config['log_file_ext'] = '.log';
`

**Comment by AlistairTB** [comment]
Finally! Some progress!

When the compose message times out, I get:

```
# cat errors.log 
[29-Aug-2021 15:31:11 -0700]: PHP Error: Connection timed out (POST /mail/?_task=mail&_unlock=loading1630276210865&_framed=1&_lang=en_US&_action=send)
[29-Aug-2021 15:31:11 -0700]: PHP Error: Failed to connect socket: Connection timed out (POST /mail/?_task=mail&_unlock=loading1630276210865&_framed=1&_lang=en_US&_action=send)
[29-Aug-2021 15:31:11 -0700]: SMTP Error: Connection failed: Failed to connect socket: Connection timed out in /var/www/html/example.com/mail/program/lib/Roundcube/rcube.php on line 1702 (POST /mail/?_task=mail&_unlock=loading1630276210865&_framed=1&_lang=en_US&_action=send)

```

**Comment by AlistairTB** [comment]
RESOLVED.

At some point between the old version I was running (1.3ish), and the 1.4.11 that was installed a few days ago, the explicit defaults in defaults.inc.php were changed so my implicit configuration from 1.3 broke.

For example, in 1.3 I did not need to specify
`
$config['smtp_user'] = '';
`
But the defaults.inc.php in version 1.4.11 made it

`
$config['smtp_user'] = '%u';
`
Mail is now working again. I'm going to go have a beer.
